File: algs/lab_1/lab_1/arraylist.cpp
#pragma once
#include "arraylist.h"

ArrayList::ArrayList(int size) {
	this->size = size;
	this->length = size * 2;
	this->array = new int*[this->length];

	if (this->size >= this->length) {
		expand();
	}

	for (int i = 0; i < size; ++i) {
		int* ptr = new int(i + 1);
		this->array[i] = ptr;
	}
}

void ArrayList::print() {
	std::cout << "[ ";

	for (int i = 0; i < this->size; ++i) {
		std::cout << *this->array[i]; 

		if (i + 1 != this->size) {
			std::cout << " , ";
		}
	}

	std::cout << " ]";
}

void ArrayList::insert(int index, int val) {
	//ÑÄÂÈÃ ÂÑÅÕ ÝËÅÌÅÍÒÎÂ

	int* tmp = NULL;
	int* prev = this->array[index];

	if (this->size + 1 >= this->length) {
		expand();
	}

	for (int i = 0; i < this->size - index; ++i) {
		tmp = this->array[index + i + 1];
		this->array[index + i + 1] = prev;
		prev = tmp;
	}

	// ÄÎÁÀÂËÅÍÈÅ ÝËÅÌÅÍÒÀ
	int* ptr = new int(val);
	this->array[index] = ptr;

	this->size += 1;
}

int ArrayList::remove(int index) {
	int* del = this->array[index];
	int val = *del;
	
	for (int i = 0; i < this->size - index; ++i) {
		this->array[index + i] = this->array[index + i + 1];
	}

	delete del;

	this->size -= 1;
	return val;
}

void ArrayList::push(int val) {
	this->insert(this->size, val);
}

int ArrayList::pop() {
	return this->remove(this->size - 1);
}

void ArrayList::push_front(int val) {
	this->insert(0, val);
}

int ArrayList::pop_front() {
	return this->remove(0);
}

void ArrayList::expand() {
	int** new_arr = new int* [this->length * 2];

	for (int i = 0; i < this->size; ++i) {
		new_arr[i] = this->array[i];
	}

	array = new_arr;
	this->length *= 2;
}

File: algs/term/structures/arraylist.py
class ArrayList:

    def __init__(self, values=None):
        if not values and values != 0:
            self.__arraylist = []
        else:
            self.__arraylist = [values]

    def get(self, index):
        return self.__arraylist[index]

    def insert(self, ind, val):
        tmp = [val]
        self.__arraylist = self.__arraylist[:ind] + tmp + self.__arraylist[ind:]

    def remove(self, ind):
        val = self.get(ind)
        self.__arraylist = self.__arraylist[:ind] + self.__arraylist[ind + 1:]
        return val

    def push(self, val):
        if len(self.__arraylist) != 0:
            self.__arraylist.insert(len(self.__arraylist), val)
        else:
            self.__arraylist.insert(0, val)

    def push_front(self, val):
        self.__arraylist.insert(0, val)

    def pop(self):
        return self.remove(len(self.__arraylist) - 1)

    def pop_front(self):
        return self.remove(0)

    def __str__(self):
        res = "<"

        for el in self.__arraylist:
            res += str(el) + ","

        if len(res) != 1:
            res = res[:-1]

        res += ">"
        return res

    def __len__(self):
        return len(self.__arraylist)

    def __getitem__(self, key):
        return self.__arraylist[key]

    def __setitem__(self, key, value):
        self.__arraylist[key] = value